compiler/rustc_ty_utils/src/nested_bodies.rs

```diff
@@ -1,15 +1,56 @@
+use rustc_data_structures::fx::FxIndexSet;
 use rustc_hir as hir;
 use rustc_hir::def_id::{DefId, LocalDefId};
 use rustc_hir::intravisit::Visitor;
 use rustc_middle::query::Providers;
+use rustc_middle::span_bug;
 use rustc_middle::ty::{self, TyCtxt};
 
 fn nested_bodies_within<'tcx>(tcx: TyCtxt<'tcx>, item: LocalDefId) -> &'tcx ty::List<LocalDefId> {
-    let body = tcx.hir_body_owned_by(item);
-    let mut collector =
-        NestedBodiesVisitor { tcx, root_def_id: item.to_def_id(), nested_bodies: vec![] };
-    collector.visit_body(body);
-    tcx.mk_local_def_ids(&collector.nested_bodies)
+    let children = if let Some(children) = tcx.resolutions(()).typeck_children.get(&item) {
+        // We re-intern this as a `List` since `List` comparisons are cheap pointer equality.
+        tcx.mk_local_def_ids_from_iter(
+            children
+                .iter()
+                .copied()
+                // We need to filter out the `Node::Err` bodies which were unsuccessfully
+                // lowered in HIR lowering, like items that are contained in never pattern
+                // match arms, which don't get lowered.
+                .filter(|def_id| !matches!(tcx.hir_node_by_def_id(*def_id), hir::Node::Err(_))),
+        )
+    } else {
+        ty::List::empty()
+    };
```

Comment:
I don't remember whether there can be typeck roots that are not hir owners. If there aren't, you could use hir::OwnerNodes::bodies.

Reply:
ohhh good point!